### PR TITLE
Yard

### PR DESCRIPTION
**Pull request by emiltin**
Using Yard to document rspec tests.
Picking up rspec specs require yard extensions, which are a bit tricky to write. So far, the generated docs will show a list of Features, but more work is needed.

### CODE REVIEW

**Comment by emiltin** [comment]
@otterdahl we could use a github action to run yardoc when pushing, and .gitignoring /doc? Otherwise we run yardoc manually and commit the docs to the repo.

**Comment by otterdahl** [comment]
> @otterdahl we could use a github action to run yardoc when pushing, and .gitignoring /doc? Otherwise we run yardoc manually and commit the docs to the repo. 

Yes. I believe it is possible. Not sure how to commit to /doc using a script
though.

**Comment by emiltin** [comment]
I don't think actions should commit to the repo. We could publish the docs somewhere, like github pages. But we would want PR branches to do that, so let's consider it a bit more before doing anything. Anyways first step is getting the yard extensions more complete.

**Comment by emiltin** [comment]
@GunnarRolander you're welcome to add commits to this branch if want to move descriptions into the code

**Comment by sweco-seguro** [comment]
Great! Do you mind if I rebase it to get the latest time synch test etc?

**Comment by emiltin** [comment]
yes you can rebase it on master

**Comment by emiltin** [comment]
I worked on the extensions and output generation.

The branch was rebased on master. If you have local commit, rebase you local yard branch on this branch, eg:
```
# on yard branch
% git fetch origin
% git rebase origin/yard
```

**Comment by emiltin** [comment]
We also need to improve out use or RSpec context, because they are used to group specifications in the documentation.

It's what you specify wit the RSpec describe() call:
```ruby
RSpec.describe 'RSMP site commands' do  
```

They can also be nested.

**Comment by emiltin** [comment]
Hi Gunnar, I think there's a problem with your rebase. The file yard/yard_extensions.rb was added, but this file was previously renamed to yard/extensions.rb. And .yardopt needs to point to the renamed file, not the old.

**Comment by emiltin** [comment]
Gunnar,
To make list works, you need to use markdown, ie. dots, not colons:
```
1. Given the site is connected
2. Request status
3. Expect status response before timeout
```

=>

1. Given the site is connected
2. Request status
3. Expect status response before timeout

I've comitted an example in the S0001 tests.

**Comment by emiltin** [comment]
I think the first step "Verify connection" would perhaps be better described as a given, e.g. "Given the site is connected", as it's really more of a preparation, that the actual test. This mirrors the language used in Cucumber.

I also think it would be nice with a line first, describing in plain text what the test does. eg:

 ```
Verify that the controller responds to S0001.
 
1. Given the site is connected
2. Request status
3. Expect status response before timeout
```

**Comment by emiltin** [comment]
Nested RSpec context are now shown in the docs. Eg this spec organization

```ruby
RSpec.describe "Traffic Light Controller" do
  include StatusHelpers
  
  describe "Signal Plans" do

    # Verify that the controller responds to S0001.
    #
    # 1. Given the site is connected
    # 2. Request status
    # 3. Expect status response before timeout

    it 'responds to S0001 signal group status', sxl: '>=1.0.7' do |example|
      request_status_and_confirm "signal group status",
        { S0001: [:signalgroupstatus, :cyclecounter, :basecyclecounter, :stage] }
    end
  end
end
```

It will be shown in the docs with nested sections.

Source code can now also be toggled on/of for specs.

**Comment by sweco-seguro** [comment]
Are all the current site tests related to Traffic Light Controllers, or are some more generic to different equipment? I.e. should all of them be nested inside a
```ruby
RSpec.describe "Traffic Light Controller" do
```

or should some, like disconnect, clock sync etc, be in some more general desciprtions?

**Comment by emiltin** [comment]
Some tests would also be relevant to other types of equipment. I think we  should place them in a RSpec context called e.g. "Core". We should probably also use folders to separate core and tlc specs:
```
spec
  site
    core
    tlc
  supervisor
```

**Comment by emiltin** [comment]
Note that all statuses and commads are we currently use are specific to traffic light controllers, since they're defined in the tlc sxl. This includes M0104 and S0096. subscribe_spec.rb also uses S0001 which is defined in the tlc sxl.
None of the tests is clock_spec.rb can therefore be said to belong to Core.

**Comment by emiltin** [comment]
@GunnarRolander let me know when you've completed import of test descriptions. Then I'll look at merging.

**Comment by sweco-seguro** [comment]
Alright, all the documentation from the test list excel is now in the code. I updated it where necessary and added some documentation to aggregated status and disconnect too.

File: spec/site/connect_spec.rb [hunk 1]
```diff
@@ -51,22 +51,52 @@ def check_sequence version
     end
   end
 
+# Verify that the handshake sequence upon connection is correct for 3.1.1
+#
+# 1. Given the site is new, connected and version 3.1.1
```
Comment (emiltin) [inline spec/site/connect_spec.rb:56]: what is the meaning of "new"?

File: spec/site/connect_spec.rb [hunk 2]
```diff
@@ -51,22 +51,52 @@ def check_sequence version
...
+# 2. Send and receive handshake messages
+# 3. Expect the handshake messages to be in the specified sequence corresponding to version 3.1.1
+# 4. Expect the site to be ready
```
Comment (emiltin) [inline spec/site/connect_spec.rb:59]: the 'site' object is a local proxy for the real equipment.
what we expect is that the site proxy object reaches the ready state. this means that the connection sequence is complete. i suggest "expect the connection sequence to be complete"

File: spec/site/connect_spec.rb [hunk 3]
```diff
@@ -51,22 +51,52 @@ def check_sequence version
...
   it 'exchanges correct connection sequence of rsmp version 3.1.1', rsmp: '3.1.1' do |example|
     check_sequence '3.1.1'
   end
 
+# Verify that the handshake sequence upon connection is correct for 3.1.2
+#
+# 1. Given the site is new, connected and version 3.1.2
```
Comment (emiltin) [inline spec/site/connect_spec.rb:66]: maybe "Verify the connection sequence when using rsmp core 3.1.2"

Comment (emiltin) [inline spec/site/connect_spec.rb:66]: maybe "Given the site is connected and using core 3.1.2"

File: spec/site/connect_spec.rb (hunk 1 above)
Comment (sweco-seguro) [reply spec/site/connect_spec.rb:56]: I was thinking of the site object, and couldn't come up with anything better at the time.

File: spec/site/connect_spec.rb (hunk 2 above)
Comment (sweco-seguro) [reply spec/site/connect_spec.rb:59]: Right, that makes sense. I'll change it.

File: spec/site/connect_spec.rb (hunk 3 above)
Comment (sweco-seguro) [reply spec/site/connect_spec.rb:66]: Sounds good, changed it.

Comment (sweco-seguro) [reply spec/site/connect_spec.rb:66]: Sounds good, changed it.

File: spec/site/connect_spec.rb [hunk 4]
```diff
@@ -1,14 +1,14 @@
 RSpec.describe "RSMP site connection" do
 
-  def check_connection_sequence version, expected
+  def check_connection_sequence version, Expected
```
Comment (emiltin) [inline spec/site/connect_spec.rb:3]: what's up with the capitalizatin here?

Comment (sweco-seguro) [reply spec/site/connect_spec.rb]: Woops, I managed to search-replace a tad more than I aimed for. Fixed it.

Comment (emiltin) [reply spec/site/connect_spec.rb]: alright :-)

**Comment by emiltin** [comment]
@GunnarRolander is this ready to merge, or do you want to move add more documentation in this first round?

**Comment by sweco-seguro** [comment]
I think it's ready to merge. Everything in the excel and some extra has been added.

**Pull request merged**